v2/apiserver/internal/core/mongodb/logs_store.go

```diff
@@ -0,0 +1,143 @@
+package mongodb
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/brigadecore/brigade/v2/apiserver/internal/core"
+	"github.com/brigadecore/brigade/v2/apiserver/internal/lib/mongodb"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// logsStore is a MongoDB-based implementation of the core.LogsStore interface.
+type logsStore struct {
+	collection mongodb.Collection
+}
+
+// NewLogsStore returns a MongoDB-based implementation of the core.LogsStore
+// interface. This implementation relies on a log aggregator having forwarded
+// and stored log entries-- a process which necessary introduces some latency.
+// Callers should favor another implementation of the core.LogsStore interface
+// and fall back on this implementation when the other fails.
+func NewLogsStore(database *mongo.Database) core.LogsStore {
+	return &logsStore{
+		collection: database.Collection("logs"),
+	}
+}
+
+func (l *logsStore) StreamLogs(
+	ctx context.Context,
+	_ core.Project,
+	event core.Event,
+	selector core.LogsSelector,
+	opts core.LogStreamOptions,
+) (<-chan core.LogEntry, error) {
+	criteria := criteriaFromSelector(event.ID, selector)
+
+	logEntryCh := make(chan core.LogEntry)
+	go func() {
+		defer close(logEntryCh)
+
+		cursorType := options.Tailable
+		var cur *mongo.Cursor
+		var err error
+		// Any attempt to open a cursor that initially retrieves nothing will yield
+		// a "dead" cursor which is no good for tailing the capped collection. We
+		// need to retry this until we get a "live" cursor or the context is
+		// canceled.
+		for {
+			cur, err = l.collection.Find(
+				ctx,
+				criteria,
+				&options.FindOptions{CursorType: &cursorType},
+			)
+			if err != nil {
+				log.Println(
+					errors.Wrap(err, "error getting cursor for logs collection"),
+				)
+				return
+			}
+			if cur.ID() != 0 {
+				// We got a live cursor.
+				break
+			}
+			if !opts.Follow {
+				// If we're not following, just return. We're done.
+				return
+			}
+			select {
+			case <-time.After(time.Second): // Wait before retry
```

Comment:
Are we missing a `continue` as is used on line 89 below?

Reply:
The end of the loop is right after this select, so as long as we fall into this case, we're about to loop around without explicitly saying `continue`.

On line 89, we're using continue to pre-empt the logic that follows the select statement and immediately begin the next iteration.

Reply:
On a separate note, however, I probably should change this to a progressive backoff. I'm not going to do that right now. I'm going to save those sort of minor optimizations for post-alpha, but I _am_ going to add a TODO comment about it.

Reply:
Ah ok, makes sense -- thanks for explaining.